File: src/hostsapi/hosts.go
// Copyright 2016 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package hello

import (
	"fmt"
	"net/http"
	"time"

	"github.com/GoogleCloudPlatform/go-endpoints/endpoints"

	"golang.org/x/net/context"
	"google.golang.org/appengine"
	"google.golang.org/appengine/datastore"
	"google.golang.org/appengine/log"
)

// Host records represent a single machine.
type Host struct {
	IPAddress string    `json:"ip_address" endpoints:"req"`
	Created   time.Time `json:"created" datastore:"created"`
}

func init() {
	// register the quotes API with cloud endpoints.
	api, err := endpoints.RegisterService(&HostsAPI{}, "hosts", "v1", "Hosts API", true)
	if err != nil {
		panic(err)
	}

	// Setup the endpoint for each API function.
	info := api.MethodByName("List").Info()
	info.Name, info.HTTPMethod, info.Path = "list", "GET", "hosts"

	info = api.MethodByName("Create").Info()
	info.Name, info.HTTPMethod, info.Path = "create", "POST", "hosts"

	info = api.MethodByName("Delete").Info()
	info.Name, info.HTTPMethod, info.Path = "delete", "POST", "hosts/{id}"

	info = api.MethodByName("Get").Info()
	info.Name, info.HTTPMethod, info.Path = "get", "GET", "hosts/{id}"

	info = api.MethodByName("Setup").Info()
	info.Name, info.HTTPMethod, info.Path = "setup", "GET", "setup"

	// Start handling cloud endpoint requests.
	endpoints.DefaultServer.ContextDecorator = AuthDecorator
	endpoints.HandleHTTP()

	// http.HandleFunc("/v1/hosts/127.0.0.1/proofOfConcept", Handler)
	http.HandleFunc("/", HelloHandler)
}

func GetHost(ctx context.Context, ipaddr string) (*Host, error) {
	var h Host
	key := datastore.NewKey(ctx, "HostsGo", ipaddr, 0, nil)
	if err := datastore.Get(ctx, key, &h); err != nil {
		return nil, err
	}
	return &h, nil
}

func PutHost(ctx context.Context, ipaddr string) (*Host, error) {
	h := Host{
		IPAddress: ipaddr,
		// NOTE: python fails to decode the format of the 'date-time' type at nanosecond resolution.
		Created: time.Now().UTC().Truncate(time.Second),
	}
	key := datastore.NewKey(ctx, "HostsGo", ipaddr, 0, nil)
	if _, err := datastore.Put(ctx, key, &h); err != nil {
		return nil, err
	}
	return &h, nil
}

func HelloHandler(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "Hello, world")
}

func Handler(w http.ResponseWriter, r *http.Request) {
	ctx := appengine.NewContext(r)

	// Get remote IP.
	remoteIp, ok := r.Header["X-Appengine-Remote-Addr"]
	if !ok {
		log.Errorf(ctx, "Found no host address for header: X-Appengine-Remote-Addr")
		return
	}
	log.Infof(ctx, "This is a test request for %s.", remoteIp[0])

	// Log all headers for fun.
	for k, v := range r.Header {
		log.Infof(ctx, "%s %s", k, v)
	}

	// Try to get a pre-defined host record.
	log.Infof(ctx, "About to call datastore.Get()")
	h, err := GetHost(ctx, remoteIp[0])
	if err != nil {
		log.Infof(ctx, "Did not find %s in datastore: %s", remoteIp[0], err)

		log.Infof(ctx, "Trying to datastore.Put() new record for: %s", remoteIp[0])
		h, err = PutHost(ctx, remoteIp[0])
		if err != nil {
			log.Infof(ctx, "Failed to put %s in datastore: %s", remoteIp[0], err)
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		log.Infof(ctx, "Saved the Host record for %#v", h)
	}
	log.Infof(ctx, "Retrieved the Host record for %#v", h)

	// Basic response.
	w.Header().Set("Content-Type", "application/json")
	fmt.Fprint(w, `{"Stage2URL": "https://storage.googleapis.com/dash-test-1/stage2/stage2.json"}`)

}

File: python/hosts.py
# Copyright 2016 Google Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""This is a sample Hosts API implemented using Google Cloud Endpoints."""


import datetime
import endpoints
import functools
import logging
import urlparse

from google.appengine.ext import ndb
from google.appengine.api import users

from endpoints_proto_datastore import ndb as endpoints_model

from protorpc import message_types
from protorpc import messages
from protorpc import remote


class Host(endpoints_model.EndpointsModel):
    """A Host Entity represents a machine record."""
    # Reset default from messages.StringField so clients see a date-time format.
    # created = endpoints_model.EndpointsDateTimeProperty() # ndb.DateTimeProperty()
    created = ndb.DateTimeProperty()

    # Should help convert DateTimeProperty to a correct date-time format, however,
    # there appears to be some bug that still prevents this:
    # https://github.com/GoogleCloudPlatform/endpoints-proto-datastore/issues/83
    # _custom_property_to_proto = {
    #     # fails.
    #     endpoints_model.EndpointsDateTimeProperty: message_types.DateTimeField
    #     # fails.
    #     ndb.DateTimeProperty: message_types.DateTimeField
    #     # works, but is the wrong type of course for a datetime object.
    #     ndb.DateTimeProperty: messages.BytesField
    # }

    # Define IdSet() and id() to require a string id.
    def IdSet(self, value):
        if not isinstance(value, basestring):
            raise TypeError('ID must be a string.')
        self.UpdateFromKey(ndb.Key(type(self), value))

    @endpoints_model.EndpointsAliasProperty(setter=IdSet, required=True)
    def id(self):
      if self.key is not None:
        return self.key.string_id()

    @property
    def ip_address(self):
        return self.id


####################################################################
# TEMPORARY
STORED_HOSTS = [
    Host(id='192.168.1.1'),
    Host(id='127.0.0.1'),
    Host(id='10.3.4.22'),
]
####################################################################


####################################################################
# The Hosts API
####################################################################

PS_CLIENT_ID = '294885104230-ue8r3k8on02m0kjsu8vh5ulnke7imnrf.apps.googleusercontent.com'
ADMIN_USERS = ['soltesz@google.com', 'stephen.soltesz@gmail.com']

ALLOWED_CLIENT_IDS = [PS_CLIENT_ID, endpoints.API_EXPLORER_CLIENT_ID]


def authorized(f):
    @functools.wraps(f)
    def admin_check(*args, **kwargs):
        user = endpoints.get_current_user()
        if not user:
            raise endpoints.ForbiddenException('Users must be authenticated!')
        if user.email() not in ADMIN_USERS:
            raise endpoints.UnauthorizedException(
                '%s is not an admin user.' % user)
        return f(*args, **kwargs)
    return admin_check


@endpoints.api(name='hosts', version='v1',
               allowed_client_ids=ALLOWED_CLIENT_IDS,
               auth_level=endpoints.AUTH_LEVEL.REQUIRED,
               description='Hosts API (Python)')
class HostsApi(remote.Service):

    # TODO: enable standard query_fields=('limit', 'order', 'pageToken').
    @Host.query_method(path='hosts', query_fields=())
    @authorized
    def list(self, query):
        logging.debug('query: %s', query)
        return query

    @Host.method(path='hosts/{id}', http_method='GET')
    @authorized
    def get(self, host):
        if not host.from_datastore:
            raise endpoints.NotFoundException(
                    'Host not found: "%s"' % host.key.id())
        return host

    @Host.method(path='hosts')
    @authorized
    def create(self, host):
        host.created = datetime.datetime.now()
        host.put()
        return host

    @Host.method(path='hosts/{id}')
    @authorized
    def delete(self, host):
        if not host.from_datastore:
            raise endpoints.NotFoundException(
                    'Host not found: "%s"' % host.key.id())
        host.key.delete()
        return host

    # TEMPORARY
    @Host.method(
        request_message=message_types.VoidMessage, path='setup',
        http_method='GET')
    @authorized
    def setup(self, unused_request):
        for host in STORED_HOSTS:
            host.created = datetime.datetime.now()
            host.put()
        return host


# API server.
api = endpoints.api_server([HostsApi])
